dask_jobqueue/core.py

```diff
@@ -1,23 +1,89 @@
+from __future__ import absolute_import, division, print_function
+
 import logging
+import math
 import os
 import shlex
 import socket
 import subprocess
 import sys
+from collections import OrderedDict
 from contextlib import contextmanager
 
 import dask
 import docrep
 from distributed import LocalCluster
 from distributed.deploy import Cluster
-from distributed.utils import get_ip_interface, ignoring, parse_bytes, tmpfile
+from distributed.utils import get_ip_interface, parse_bytes, tmpfile
+from distributed.diagnostics.plugin import SchedulerPlugin
+from sortedcontainers import SortedDict
 
 dirname = os.path.dirname(sys.executable)
 
 logger = logging.getLogger(__name__)
 docstrings = docrep.DocstringProcessor()
 
 
+def _job_id_from_worker_name(name):
+    ''' utility to parse the job ID from the worker name'''
+    return name.split('-')[-2]
+
+
+class Job(object):
+    def __init__(self, job_id, status=None):
+        self.job_id = job_id
+        self.status = status
+        self.workers = SortedDict()
+
+    def update(self, worker=None, status=None):
+        ''' update the status this job'''
+        if worker is not None:
+            self.workers[worker.address] = worker
+        if status is not None:
+            self.status = status
+
+    def __repr__(self):
+        return "<%s: %r, status: %r, workers: %r>" % (
+            self.__class__.__name__, self.job_id, self.status, self.workers)
+
+
+class JobQueuePlugin(SchedulerPlugin):
+    def __init__(self):
+        self.pending_jobs = OrderedDict()
+        self.running_jobs = OrderedDict()
+        self.finished_jobs = OrderedDict()
+
+    def add_worker(self, scheduler, worker=None, name=None, **kwargs):
+        ''' Run when a new worker enters the cluster'''
+        w = scheduler.workers[worker]
+        job_id = _job_id_from_worker_name(w.name)
+
+        # if this is the first worker for this job, move job to running
+        if job_id not in self.running_jobs:
+            self.running_jobs[job_id] = self.pending_jobs.pop(job_id)
+            self.running_jobs[job_id].update(status='running')
+
+        # add worker to dict of workers in this job
+        self.running_jobs[job_id].update(worker=w)
+
+    def remove_worker(self, scheduler=None, worker=None, **kwargs):
+        ''' Run when a worker leaves the cluster'''
+        # the worker may have already been removed from the scheduler so we
+        # need to check in running_jobs for a job that has this worker
+        for job_id, job in self.running_jobs.items():
+            if worker in job.workers:
+                # remove worker from dict of workers on this job id
+                del self.running_jobs[job_id].workers[worker]
+                break
+        else:
+            raise ValueError('did not find a job that owned this worker')
```

Comment:
This will also run if a worker just restarts, or is temporarily killed by a nanny.  We might not want to remove the job entirely here.

Reply:
Ah, I see that we just remove one of the workers from the job.  I guess we add the worker back in when it starts back up in the `add_worker` function?